nhrpd/netlink_arp.c

```diff
@@ -93,20 +97,33 @@ static void netlink_neigh_msg(struct nlmsghdr *msg, struct zbuf *zb)
 	if (!c)
 		return;
 
-	if (msg->nlmsg_type == RTM_GETNEIGH) {
-		debugf(NHRP_DEBUG_KERNEL, "Netlink: who-has %s dev %s",
-		       sockunion2str(&addr, buf, sizeof buf), ifp->name);
+	debugf(NHRP_DEBUG_KERNEL,
+	       "Netlink: %s %s dev %s lladdr %s nud 0x%x "
+	       "cache used %u type %u",
+	       (msg->nlmsg_type == RTM_GETNEIGH)
+		       ? "who-has"
+		       : (msg->nlmsg_type == RTM_NEWNEIGH) ? "new-nei"
+							   : "del-nei",
+	       sockunion2str(&addr, buf[0], sizeof(buf[0])), ifp->name,
+	       sockunion2str(&lladdr, buf[1], sizeof(buf[1])), ndm->ndm_state,
+	       c->used, c->cur.type);
 
+	if (msg->nlmsg_type == RTM_GETNEIGH) {
 		if (c->cur.type >= NHRP_CACHE_CACHED) {
 			nhrp_cache_set_used(c, 1);
-			netlink_update_binding(ifp, &addr,
-					       &c->cur.peer->vc->remote.nbma);
+			debugf(NHRP_DEBUG_KERNEL,
+			       "Netlink: update binding for %s dev %s from c "
+			       "%s peer.vc.nbma %s to lladdr %s",
+			       sockunion2str(&addr, buf[0], sizeof(buf[0])),
+			       ifp->name,
+			       sockunion2str(&c->cur.remote_nbma_natoa, buf[1],
+					     sizeof(buf[1])),
+			       sockunion2str(&c->cur.peer->vc->remote.nbma,
+					     buf[2], sizeof(buf[2])),
+			       sockunion2str(&lladdr, buf[3], sizeof(buf[3])));
+			netlink_update_binding(ifp, &addr, &lladdr);
 		}
```

Comment:
can you elaborate why 3rd parameter is lladdr instead of c->cur.peer->vc->remote.nbma ?

Reply:
If the cached entry is a dynamically-cached NHRP Shortcut (i.e. spoke-to-spoke), then `lladdr` gives its correct nbma address , which is different from  vc->remote.nbma (nhs' nbma address).